docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

```diff
@@ -0,0 +1,1680 @@
+variable of the same name.
+
+.. _example-1:
+
+**Example**
+
+Below, the parameterized data type on the right is a valid upgrade of
+the parameterized data type on the left. As is valid with any record
+type, it adds an optional field.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+            data Tree a = 
+              Tree with 
+                label : a
+                children : [Tree a]
+
+     - .. code-block:: daml
+
+            data Tree a = 
+              Tree with 
+                label : a
+                children : [Tree a]
+                cachedSize : Optional Int
+
+Below, the parameterized data type on the right is **not** a valid
+upgrade of the parameterized data type on the left. It renames the type
+variable ``a`` into ``b`` which is currently not supported.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+            data Tree b = 
+              Tree with 
+                label : b
+                children : [Tree b]
+
+     - .. code-block:: daml
+
+            data Tree b = 
+              Tree with 
+                label : b
+                children : [Tree b]
+
+Interfaces
+~~~~~~~~~~
+
+Interfaces cannot be upgraded. We strongly discourage
+uploading a package that contains both interface and template
+definitions, as these templates cannot benefit from smart contract upgrades
+in the future. Instead, we recommend declaring interfaces in a
+package of their own that defines no template.
+
+Data Transformation: Runtime Semantics
+--------------------------------------
+
+Whenever a contract is fetched or one of its choices is exercised, it is
+retrieved from the ledger and transformed into a value that fits the target
+template type. Then, its metadata (signatories, observers, key, maintainers) is
+recomputed using the code of the target temmpate and compared against the
+existing metadata: it is not allowed to change. The ensure clause of the 
+contract is also re-evaluated.
+
+In addition, when a choice is exercised, its arguments are transformed into
+values that fit the type of the parameters of the choice in the target package.
+The result of the exercise is then possibly transformed back to some other
+target type by the client (e.g. the generated java client code).
+
+Below, we detail the rules governing target types, then explain how
+transformations are performed, and finally detail the rules of metadata
+re-computation.
+
+Target Types
+~~~~~~~~~~~~
+
+In a top-level fetch or exercise triggered by a Ledger API command, the
+target template type is determined by the rules of package preference
+detailed in the `Ledger API <#_swcg5hoxdia8>`__ section. Once a given
+version of a template has been selected, the target type of its
+parameters as well as the target type of the contracts retrieved in its
+choices is determined by their static type.
```

Comment:
Unclear at his point what you mean by "their static type".

Reply:
I mean that the type is determined at compile time. I'll try to find a better way to formulate it.

Reply:
Clarified.